### PR TITLE
Cache pages

### PR DESCRIPTION
**Pull request by tsauerwein**
This PR enables page caching. For example when requesting a document `routes/12345/fr`, the ui server side will check if there is an entry in the cache. If so, a request to the api is made with the previous ETag value (which is stored with the cached page). If the document on the api side has not changed, the cached page is returned. Otherwise the page is rendered, and the cached entry is updated.
An ETag is header is set, which is used by the browser cache (and also Varnish). The ETag header consists of the ETag from the api and the current Git revision (+ a timestamp when running the dev server).

Todo:
* [x] Set up caching/ETag for
  * [x] Detail page
  * [x] Archive page
  * [x] History page
  * [x] Diff page
  * [x] Index page
  * [x] Add/~~edit~~ page
  * [x] Index/auth/account page

Depends on https://github.com/c2corg/v6_api/pull/336
Related to https://github.com/c2corg/v6_ui/issues/456

### CODE REVIEW

**Comment by tsauerwein** [comment]
This PR would be ready for review.

I didn't set up caching for editing pages (e.g. `/waypoints/edit/733087/fr`). Edit pages for different documents differ in two lines (the lines where the document id and lang are set). So, a separate page for each document and lang would have to be cached. Because there are less edits than views, and to avoid that the cache is filled up too much, I decided no to cache these pages.

File: c2corg_ui/__init__.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
 import requests
 from pyramid.config import Configurator
-from pyramid_mako import add_mako_renderer
-from c2corg_ui.lib.cacheversion import version_cache_buster, CACHE_PATH
 from pyramid.httpexceptions import (
     HTTPBadRequest, HTTPInternalServerError)
-from pyramid.view import view_config
 from pyramid.view import notfound_view_config
+from pyramid.view import view_config
+from pyramid_mako import add_mako_renderer
```
Comment (asaunier) [inline c2corg_ui/__init__.py]: Is this still needed?

**Comment by asaunier** [comment]
> I didn't set up caching for editing pages
> Because there are less edits than views, and to avoid that the cache is filled up too much, I decided not to cache these pages.

OK that makes sense. The most important is that detail view pages and very common pages (homepage, advanced search/listing pages if possible) are cached. History and archived versions caching is a plus.

If I understand well, pages are cached a several levels:
* on the API side
* in dogpile for the UI server side
* in Varnish and in the user's browser
Right?

Etags are propagated from the browser to the API to figure out if a new version of the page is available. If yes, the various cache levels are refreshed. Correct?

File: c2corg_ui/__init__.py (hunk 1 above)
Comment (tsauerwein) [inline c2corg_ui/__init__.py]: Yes, the pages are still rendered with Mako.

**Comment by tsauerwein** [comment]
> If I understand well, pages are cached a several levels:

Yes, there will also be a Varnish for the API.

> Etags are propagated from the browser to the API to figure out if a new version of the page is available. If yes, the various cache levels are refreshed. Correct?

Yes. The UI server side also stores the ETag together with the rendered page. For example, if a browser requests a page without providing an ETag, but there is a cache entry for the page, the request to the API will include the ETag from the cache (to check if the entry in the cache is still up-to-date).

**Comment by tsauerwein** [comment]
@asaunier Are you ok with merging this one?

**Comment by asaunier** [comment]
Please do! :+1:

**Comment by tsauerwein** [comment]
Thanks!

**Pull request merged**